File: backend/java/src/main/java/me/lukasgrimm/bachelor/dao/PostDBDAO.java
package me.lukasgrimm.bachelor.dao;

import me.lukasgrimm.bachelor.models.Post;

import javax.persistence.EntityManager;
import javax.persistence.EntityManagerFactory;
import javax.persistence.PersistenceException;
import javax.persistence.Query;
import java.util.List;

public class PostDBDAO implements PostDAO {

    private EntityManagerFactory entityManagerFactory;

    public PostDBDAO(EntityManagerFactory entityManagerFactory) {
        this.entityManagerFactory = entityManagerFactory;
    }

    public List<Post> findAll() {

        EntityManager entityManager = null;
        try {
            entityManager = entityManagerFactory.createEntityManager();
            Query query = entityManager.createQuery("SELECT p FROM Post p");
            List<Post> list = query.getResultList();
            list.sort((postA, postB) -> (int)(postA.getId() - postB.getId()));
            return list;
        } finally {
            if (entityManager != null) {
                entityManager.close();
            }
        }
    }

    public void create(Post post) {

        EntityManager entityManager = null;
        try {
            entityManager = entityManagerFactory.createEntityManager();
            entityManager.getTransaction().begin();
            entityManager.persist(post);
            entityManager.getTransaction().commit();
        } catch (PersistenceException e) {
            throw new IllegalArgumentException("id may not be specified");
        } finally {
            if (entityManager != null) {
                entityManager.close();
            }
        }
    }

    public void update(Post post) {

        EntityManager entityManager = null;
        try {
            entityManager = entityManagerFactory.createEntityManager();

            if (entityManager.find(Post.class, post.getId()) == null) {
                throw new IllegalArgumentException("no post found");
            }

            entityManager.getTransaction().begin();
            entityManager.merge(post);
            entityManager.getTransaction().commit();
        } finally {
            if (entityManager != null) {
                entityManager.close();
            }
        }
    }

    public void delete(long postId) {

        EntityManager entityManager = null;
        try {
            entityManager = entityManagerFactory.createEntityManager();
            entityManager.getTransaction().begin();
            Post post = entityManager.find(Post.class, postId);

            if (post == null) {
                throw new IllegalArgumentException("post not found");
            }

            entityManager.remove(post);
            entityManager.getTransaction().commit();
        } finally {
            if (entityManager != null) {
                entityManager.close();
            }
        }
    }
}

File: backend/php/dao/src/dao/PostDBDAO.php
<?php

namespace dao;

use Doctrine\ORM\EntityManager;
use model\Post;
use dao\PostDAO;

class PostDBDAO implements PostDAO
{

    private $entityManager;

    function __construct(EntityManager $entityManager)
    {
        $this->entityManager = $entityManager;
    }

    static function compareById($a, $b)
    {
        return $a->getId() - $b->getId();
    }

    function findAll()
    {
        $postRepository = $this->entityManager->getRepository('\model\Post');
        $list = $postRepository->findAll();
        usort($list, '\dao\PostDBDAO::compareById');
        return $list;
    }

    function create(Post $post)
    {
        if ($post->getId()) {
            throw new \exception\IllegalArgumentException();
        }

        $this->entityManager->persist($post);
        $this->entityManager->flush();
    }

    function update(Post $post)
    {
        $managedPost = $this->entityManager->find('\model\Post', $post->getId());

        if (!$managedPost) {
            throw new \exception\IllegalArgumentException();
        }

        $this->entityManager->merge($post);
        $this->entityManager->flush();
    }

    function delete(int $id)
    {
        $managedPost = $this->entityManager->find('\model\Post', $id);

        if (!$managedPost) {
            throw new \exception\IllegalArgumentException();
        }

        $this->entityManager->remove($managedPost);
        $this->entityManager->flush();
    }
}